cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -123,7 +123,15 @@ func NewKubeStaticPodPathManagerUsingTempDirs(kubernetesDir, patchesDir string,
 
 // MoveFile should move a file from oldPath to newPath
 func (spm *KubeStaticPodPathManager) MoveFile(oldPath, newPath string) error {
-	return os.Rename(oldPath, newPath)
+	err := os.Rename(oldPath, newPath)
+	if err != nil {
+		klog.V(4).Infof("failed to rename %v to %v: %v, fallback to copy file", oldPath, newPath, err)
+		if err := kubeadmutil.CopyFile(oldPath, newPath); err != nil {
```

Comment:
is it safe to use it directly instead of a fallback? seems better to me.

Reply:
`os.Rename` has better performance, this is useful for copying large files

Reply:
> klog.V(4).Infof("failed to rename %v to %v: %v, fallback to copy file", oldPath, newPath, err)

users might find this message confusing.

suggesting:
> klog.V(4).Infof("failed to rename %v to %v: %v, attempting an alternative method", oldPath, newPath, err)

Reply:
can we do better error handling here? 

I mean we should only call this alternative function for copying the file across the devices if we see the cross device link error

also, can we put some comments to describe the cross device link error we saw and mention how this function solves that problem

Reply:
the error checking can be done like so:
https://gist.github.com/var23rav/23ae5d0d4d830aff886c3c970b8f6c6b?permalink_comment_id=4431960#gistcomment-4431960

```
$ grep "os.Rename" * -rnI
app/phases/upgrade/postupgrade_test.go:215:		if err := os.Rename(from, to); err != nil {
app/phases/upgrade/postupgrade_test.go:227:		if err := os.Rename(from, to); err != nil {
app/phases/upgrade/staticpods_test.go:327:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods_test.go:338:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods_test.go:349:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods_test.go:360:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods_test.go:376:				return os.Rename(oldPath, newPath)
app/phases/upgrade/staticpods_test.go:393:				return os.Rename(oldPath, newPath)
app/phases/upgrade/staticpods_test.go:410:				return os.Rename(oldPath, newPath)
app/phases/upgrade/staticpods_test.go:422:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods_test.go:434:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods_test.go:446:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods_test.go:457:			moveFileFunc:         os.Rename,
app/phases/upgrade/staticpods.go:126:	return os.Rename(oldPath, newPath)

```

so we seem to not have any other locations with os.Rename outside of tests, but i think we should:
- create a new function kubeadmutil.MoveFile()
- use this function in staticpods.go
- it tries to do the os.Rename if it fails with the cross link error use the alternative way CopyFile / Remove
- add some comments why this can happen

Reply:
Updated